### PR TITLE
feat(indexRootPaths): invalidate index.html root paths

### LINKED ISSUE #9: Invalidate root paths for index.html objects
When using CF in front of an S3 website, you can configure it such that `/foo/` resolves to `/foo/index.html`.  Currently, this plugin only invalidates the `/foo/index.html` object.  When users visit `/foo/` the old index.html is still served.  This is because the `/foo/` path is it's own object and also needs invalidated.

You can see this same issue outlined and fixed in the `s3_website` gem sometime ago [here](https://github.com/laurilehmijoki/s3_website/issues/29).

The solution is to invalidate the root path for an index.html if the `file.s3.path` is an index.html.  As with `s3_website` it would probably make sense to put this behind an option and document it.

If this is something you would merge, I'll make PR.  Please let me know.

### PR DESCRIPTION
**Pull request by levithomason**
~~Haven't tested this yet, will tomorrow during the work day.~~ Works, see below 🐱 

Fixes #9

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,7 +43,8 @@ var cfSettings = {
   accessKeyId: '...',             // Optional AWS Access Key ID
   secretAccessKey: '...',         // Optional AWS Secret Access Key
   sessionToken: '...',            // Optional AWS Session Token
-  wait: true                      // Whether to wait until invalidation is completed (default: false)
+  wait: true,                     // Whether to wait until invalidation is completed (default: false)
+  indexRootPaths: true            // Invalidate index.html root paths (`foo/index.html` and `foo/`) (default: false)
```
Comment (levithomason) [inline README.md:47]: Documented the option.

File: index.js [hunk 2]
```diff
@@ -3,7 +3,8 @@ var util = require('gulp-util')
   , aws = require('aws-sdk');
 
 module.exports = function (options) {
-  options.wait = (options.wait === undefined || !options.wait) ? false : true;
+  options.wait = !!options.wait;
+  options.indexRoot = !!options.indexRoot;
```
Comment (levithomason) [inline index.js:7]: Defaults to false.  No need to use a conditional only to return a boolean if/else.  Just return the conditional.

File: index.js [hunk 3]
```diff
@@ -16,8 +17,8 @@ module.exports = function (options) {
   var files = [];
 
   var complain = function (err, msg, callback) {
+    callback(false);
     throw new util.PluginError('gulp-cloudfront-invalidate', msg + ': ' + err);
-    return callback(false);
```
Comment (levithomason) [inline index.js:20]: This was unreachable code.  The throw terminates the program.  I moved the callback above the throw.

File: index.js [hunk 4]
```diff
@@ -51,6 +52,9 @@ module.exports = function (options) {
       case 'create':
       case 'delete':
         files.push(file.s3.path);
+        if (options.indexRootPath && /index\.html$/.test(file.s3.path)) {
+          files.push(file.s3.path.replace(/index\.html$/, ''))
+        }
```
Comment (levithomason) [inline index.js:57]: Here is the point of this PR, testing if the S3 path ends with `index.html`, if so, we push another file that just replaces the `index.html` with nothing.

**Comment by levithomason** [comment]
@lpender This PR seems to be working beautifully for us.  Here is a my test case:

**Gulp Task**

Note `indexRootPath: true`

``` js
gulp.task('publish:aws', (cb) => {
  const publishHeaders = {
    'Cache-Control': 'max-age=315360000, no-transform, public',
    'x-amz-acl': 'public-read',
  }

  const s3Opts = {
    accessKeyId,
    secretAccessKey,
    params: {
      Bucket: paths.publish.s3Bucket,
    },
  }

  const cloudfrontOpts = {
    accessKeyId,
    secretAccessKey,
    distribution,
    indexRootPath: true,       // <-- Here's the new option
  }

  const publisher = g.awspublish.create(s3Opts, {
    cacheFileName: '.awspublish-cache'
  })

  return gulp.src(paths.publish.s3Src)
    .pipe(awspublish.gzip())
    .pipe(parallelize(publisher.publish(publishHeaders), 10))
    .pipe(publisher.sync())
    .pipe(cloudfrontInvalidateAwsPublish(cloudfrontOpts))
    .pipe(publisher.cache())
    .pipe(awspublish.reporter())
})
```

**CloudFront Invalidation**

Notice all the `index.html` root paths were invalidated.  The `.xml` files can be seen as not invalidating their root path `/`.

```
/about-us/
/about-us/index.html
/careers/
/careers/index.html
/contact-us/
/contact-us/index.html
/feed-50d794812f.xml
/feed-bc5f5fe623.xml
/press/
/press/index.html
/press/example-post/
/press/example-post/index.html
/press/releases/
/press/releases/index.html
/press/releases/example-press-release/
/press/releases/example-press-release/index.html
/press/releases/hot-off-the-press/
/press/releases/hot-off-the-press/index.html
/press/resources/
/press/resources/index.html
...
```

**Comment by levithomason** [comment]
Bump @lpender

**Comment by lpender** [comment]
Thanks for the PR

**Comment by lpender** [comment]
And the careful testing/documenting

**Comment by lpender** [comment]
Published 0.2.0

**Pull request merged**